### PR TITLE
[SPARK-24557][ML] ClusteringEvaluator support array input

### PR DESCRIPTION
**Pull request by zhengruifeng**
## What changes were proposed in this pull request?
ClusteringEvaluator support array input

## How was this patch tested?
added tests

### CODE REVIEW

File: mllib/src/main/scala/org/apache/spark/ml/evaluation/ClusteringEvaluator.scala [hunk 1]
```diff
@@ -107,15 +106,18 @@ class ClusteringEvaluator @Since("2.3.0") (@Since("2.3.0") override val uid: Str
 
   @Since("2.3.0")
   override def evaluate(dataset: Dataset[_]): Double = {
-    SchemaUtils.checkColumnType(dataset.schema, $(featuresCol), new VectorUDT)
+    SchemaUtils.validateVectorCompatibleColumn(dataset.schema, $(featuresCol))
     SchemaUtils.checkNumericType(dataset.schema, $(predictionCol))
 
+    val vectorCol = DatasetUtils.columnToVector(dataset, $(featuresCol))
+    val df = dataset.select(col($(predictionCol)),
```
Comment (mgaido91) [inline mllib/src/main/scala/org/apache/spark/ml/evaluation/ClusteringEvaluator.scala]: not sure this is the right way. Probably we can face the same issue everywhere we are using `DatasetUtils.columnToVector`. Probably it is better to fix the problem there.

Comment (zhengruifeng) [reply mllib/src/main/scala/org/apache/spark/ml/evaluation/ClusteringEvaluator.scala]: @mgaido91  Thanks for your reviewing!
I have considered this, however there exists a problem:
if we want to append metadata into the transformed column (like using method `.as(alias: String, metadata: Metadata)`) in `DatasetUtils.columnToVector`, how can we get the name of transformed column?
The only way to do this I know is:
```
val metadata = ...
val vectorCol = ..
val vectorName = dataset.select(vectorCol) .schema.head.name
vectorCol.as(vectorName, metadata)
```

Comment (mgaido91) [reply mllib/src/main/scala/org/apache/spark/ml/evaluation/ClusteringEvaluator.scala]: we have the new column we are returning, so we can easily get its name with `.name`

Comment (zhengruifeng) [reply mllib/src/main/scala/org/apache/spark/ml/evaluation/ClusteringEvaluator.scala]: @mgaido91  I think it maybe nice to first add a name getter for column

Comment (mgaido91) [reply mllib/src/main/scala/org/apache/spark/ml/evaluation/ClusteringEvaluator.scala]: we can propose that

**Comment by zhengruifeng** [comment]
@mgaido91 I am sorry to make a force push to update my git username in this PR. 
Since I found that my current PRs are not linked to my account and it is troublesome to track them.

and I found that `.expr.sql` may not be a good way to get the output column name:
```
val df = Seq((1,2),(3,4),(1,1),(2,-1),(3,1)).toDF("a","b")
val z = (col("a") + col("b")).as("x").as("z")
scala> z.expr.sql
res15: String = (`a` + `b`) AS `x` AS `z`
```
So I still keep the orignal method.

@MLnick @mengxr @jkbradley Would you please help reveiwing this? Thanks!

**Comment by zhengruifeng** [comment]
retest this please

**Comment by zhengruifeng** [comment]
@mengxr I notice that you open a ticket for supporting integer type labels in ClusteringEvalutator, would you like to shepherd this pr too?

**Comment by mengxr** [comment]
LGTM. Merged into master. Thanks!

**Pull request closed without merging**